### ISSUE TITLE
extract check_present and order_matters for generic comparisons

### ISSUE THREAD

**Issue by graingert** [Simplistix/testfixtures#148]
I'm trying to test some events that occur in a system similar to LogCapture, I'd like to be able to compare sequences with these flags sometimes:

```
compare_sequence(actual, expected, order_matters=True, check_present=True)
```

**Comment by cjw296** [comment]
It would be interesting to see what a PR for this looked like. I'm +0.5 on the idea... often want unordered myself, but not that I couldn't just cast to a set.

Partial matching implied by `check_present` is going to be difficult, eg:
```
compare(expected=[{1: 2}, {3: 4}], actual=[{1:2), {3:5}, {4: 4}])
```

Does that raise an exception or not?

**Comment by asqui** [comment]
@graingert : Do you have some examples of where you would like this functionality?

I usually achieve unordered comparisons by just constructing a `set` on both sides.
The check_present scenario is a bit more interesting.
In some cases we want to assert on subsets of large dictionaries so I created a `DictSubsetComparison` object that you could use as your `expected` and have it only compare the keys in that.

**Comment by graingert** [comment]
```
compare_sequence(expected=[{1: 2}, {3: 4}], actual=[{1:2}, {3:5}, {4: 4}], order_matters=True, check_present=True)
```
I'd expect to assert because the actual value is missing `{3: 4}`
```

**Comment by graingert** [comment]
> Do you have some examples of where you would like this functionality?

I have tests for something that dispatches events (like log messages) and I want to know that all the events I'm interested in get sent in the right order

**Comment by cjw296** [comment]
> ```
> compare_sequence(expected=[{1: 2}, {3: 4}], actual=[{1:2}, {3:5}, {4: 4}], order_matters=True, check_present=True)
> ```
> 
> I'd expect to assert because the actual value is missing `{3: 4}`

Really? 'cos it's there, it just maps to 5 not 4... ;-)

@asqui 's right though, this might well be better suited to a comparison object, some examples on existing one:

https://testfixtures.readthedocs.io/en/latest/comparing.html#round-comparison-objects
https://testfixtures.readthedocs.io/en/latest/comparing.html#range-comparison-objects
https://testfixtures.readthedocs.io/en/latest/comparing.html#string-comparison-objects

This will give you much more explicit control over where and how you want sequences to be considered unordered or partially matching.

@asqui - can't remember why `DictSubsetComparison` (other than the not-very-snappy name) didn't end up in `testfixtures` itself, was I having a grumpy moment?

**Comment by graingert** [comment]
> eally? 'cos it's there, it just maps to 5 not 4..

By which I mean I don't see it useful that this feature would apply recursively

**Comment by cjw296** [comment]
Right, but comparers are always used in recursive contexts with nested data structures, that's the intention of their existence.
Comparison objects are designed for explicit use, and I think they're a better fit here.
I'd certainly welcome an `UnorderedSequence` comparer and a `PartialSequence` comparer, or a single one with options that you could bake as follows:

```python

S = partial(SequenceComparison, order_matters=True, partial=True)

def test_something():
    compare(whatever, expected={'key': SequenceComparison('item1', 'item2')})`
```

@asqui - how would you feel back making a similar `DictComparison` object, sketch:

```
class DictComparison(dict):

    def __init__(*args, partial=False, order_matters=(True if PY3 else False), **kw):
        if PY2 and order_matters:
            raise TypeError('dicts are unordered on Python 2')
        super.__init__(*args, **kw)
        self.partial=True
        self.order_matters=True
````

This would cover your original use use case, as well as allowing easy checking on insertion order on Py3.
Maybe Py2 could use OrderedDict instead or something?

**Comment by graingert** [comment]
UnorderedSequence and PartialSequence makes sense to me

**Comment by cjw296** [comment]
Cool, although the more I think about it, `S = partial(SequenceComparison, order_matters=True, partial=True)` is probably preferable, otherwise you need then also need an UnorderedPartialSequence... Thoughts?

**Comment by graingert** [comment]
UnorderedPartialSequence is just a Subset
UnorderedSequence would be better named `Permutation`

**Comment by cjw296** [comment]
How about:

```python

class SequenceComparison(...):
   def __init__(self, ordered=True, partial=False):
   ...

Subset = partial(SequenceComparison, ordered=False, partial=True)
Permutation = partial(SequenceComparison, ordered=False)
```

**Comment by graingert** [comment]
sounds really great

**Comment by cjw296** [comment]
Well, that was a fair bit more work than expected, but done now in 18c336028b2786706573e8013499f02d8cf025e7 and 2648c5a332987102713206758763d16ee4d5c663.

**Comment by graingert** [comment]
Amazing thank you!

**Comment by graingert** [comment]
@cjw296 can you make a release on pypi with this feature, please?

**Comment by cjw296** [comment]
Released in 6.16.0.
